File: server/users/services/google.py
"""
Google OAuth 검증 서비스

Google OAuth2 API를 사용하여 액세스 토큰을 검증하고 사용자 정보를 가져옵니다.
공식 문서: https://developers.google.com/identity/protocols/oauth2
"""
import httpx
import logging
from typing import Optional, Dict, Any
from django.conf import settings

logger = logging.getLogger(__name__)


async def verify_google_token(access_token: str) -> Optional[Dict[str, Any]]:
    """
    Google 액세스 토큰 검증 및 사용자 정보 조회

    Args:
        access_token: Google 액세스 토큰 (Flutter SDK로부터 받은 토큰)

    Returns:
        dict: 사용자 정보 딕셔너리
            {
                'id': str,  # Google 사용자 ID (sub)
                'email': str,  # 이메일
                'name': str,  # 이름
                'profile_image': str,  # 프로필 이미지 URL
            }
        None: 검증 실패 시
    """
    try:
        async with httpx.AsyncClient() as client:
            # Google UserInfo API
            # https://www.googleapis.com/oauth2/v2/userinfo
            response = await client.get(
                'https://www.googleapis.com/oauth2/v2/userinfo',
                headers={
                    'Authorization': f'Bearer {access_token}',
                }
            )

            if response.status_code != 200:
                # 토큰이 유효하지 않거나 만료됨
                return None

            data = response.json()

            # 사용자 정보 추출
            user_info = {
                'id': data.get('id'),  # Google 사용자 ID (필수)
                'email': data.get('email', ''),  # 이메일
                'name': data.get('name', ''),  # 이름
                'profile_image': data.get('picture', ''),  # 프로필 이미지
            }

            # 필수 필드 확인
            if not user_info['id']:
                return None

            # 이메일 검증 여부 확인
            if not data.get('verified_email', False):
                # 이메일이 검증되지 않은 경우
                return None

            return user_info

    except Exception as e:
        # 네트워크 에러, JSON 파싱 에러 등
        logger.error(f"Google token verification error: {e}")
        return None


async def get_google_user_info(access_token: str) -> Optional[Dict[str, Any]]:
    """
    Google 사용자 정보 조회 (verify_google_token의 alias)

    Args:
        access_token: Google 액세스 토큰

    Returns:
        dict: 사용자 정보 또는 None
    """
    return await verify_google_token(access_token)

File: server/pkg/sns/google.go
package sns

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// GoogleUserInfo represents Google user information
type GoogleUserInfo struct {
	ID           string `json:"id"`
	Email        string `json:"email"`
	Name         string `json:"name"`
	ProfileImage string `json:"profile_image"`
}

// GoogleAPIResponse represents Google UserInfo API response
type GoogleAPIResponse struct {
	ID            string `json:"id"`
	Email         string `json:"email"`
	VerifiedEmail bool   `json:"verified_email"`
	Name          string `json:"name"`
	GivenName     string `json:"given_name"`
	FamilyName    string `json:"family_name"`
	Picture       string `json:"picture"`
}

// VerifyGoogleToken verifies Google access token and returns user info
func VerifyGoogleToken(ctx context.Context, accessToken string) (*GoogleUserInfo, error) {
	client := &http.Client{Timeout: 10 * time.Second}

	req, err := http.NewRequestWithContext(ctx, "GET", "https://www.googleapis.com/oauth2/v2/userinfo", nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+accessToken)

	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("google API returned status %d", resp.StatusCode)
	}

	var googleResp GoogleAPIResponse
	if err := json.NewDecoder(resp.Body).Decode(&googleResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	if googleResp.ID == "" {
		return nil, fmt.Errorf("invalid google user id")
	}

	if !googleResp.VerifiedEmail {
		return nil, fmt.Errorf("email not verified")
	}

	return &GoogleUserInfo{
		ID:           googleResp.ID,
		Email:        googleResp.Email,
		Name:         googleResp.Name,
		ProfileImage: googleResp.Picture,
	}, nil
}
